### PR TITLE
Migrating Marek's code for DBI to discontinue the old repository

### PR DESCRIPTION
**Pull request by Sam-XiaoyueLi**
Checklist:
- [ ] Reviewers confirm new code works as expected.
- [x] Tests are passing.
- [x] Coverage does not decrease.
- [x] Documentation is updated.

Hi guys, this is our first pull request. @marekgluza  and I have reviewed the code together and to the best of our knowledge it is ready for your review.  The changes include:
1. New file: `additional_double_bracket_functions.py` includes additional functions, among which the most important ones are `select_best_dbr_generator` and `select_best_dbr_generator_and_run`. You may see how it works in the example notebook "DBI_strategies_Pauli-Z_products.ipynb".
2. To ensure the functions running, it was necessary to make some small changes to the "double_bracket.py" file, you may verify that the changes do not affect the normal routine as in "dbi.ipynb" (now renamed "dbi_tutotrial_basic_intro.ipynb"

In summary, you can try running the "DBI_strategies_Pauli-Z_products.ipynb". This will show you which functions are being used.

### CODE REVIEW

**Review by andrea-pasquale — CHANGES_REQUESTED** [review]
Thanks @Sam-XiaoyueLi for adding this interesting features.
You can find below some comments. Let me know if you need help with any of the points that I raised.

File: src/qibo/models/dbi/utils.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+        # print(f'canonical step {step}, loss {dbi_object.off_diagonal_norm}')
+        dbi_object.h = deepcopy(h_before)
+        dbi_object.mode = generator_type
+    # find best d
+    # print(norms_off_diagonal_restriction)
+    idx_max_loss = norms_off_diagonal_restriction.index(
+        min(norms_off_diagonal_restriction)
+    )
+    step_optimal = optimal_steps[idx_max_loss]
+    return idx_max_loss, step_optimal
```
Comment (Edoardo-Pedicillo) [inline src/qibo/models/dbi/utils.py:127]: I suggest returning the `dbi_object` with the best mode, `step_optimal` and `norms_off_diagonal_restriction`, in this way the user can run the best dbi autonomously after calling this function, i.e. we can delete `select_best_dbr_generator_and_run`.

File: src/qibo/models/dbi/utils.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+
+
+def str_to_op(name: str):
+    tensor_op = 1
+    for qubit, char in enumerate(name):
+        tensor_op *= getattr(symbols, char)(qubit)
+    return tensor_op
+
+
+def select_best_dbr_generator(
```
[the remark this answers is not in the document]
Comment (Sam-XiaoyueLi) [reply src/qibo/models/dbi/utils.py:61]: double bracket rotation, this name is suggested by @marekgluza

File: src/qibo/models/dbi/utils.py (hunk 1 above)
Comment (Sam-XiaoyueLi) [reply src/qibo/models/dbi/utils.py:127]: Please correct me if I understood wrongly... so we deepcopy `dbi_object` in the function and update it by default? To run the double bracket rotation, the user will do `dbi_input, idx_max_loss, step_optimal = select_best_dbr_generator(dbi_input)`?

File: src/qibo/models/dbi/utils.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+        if "Z" in op_string_tup:
+            op_name = "".join(op_string_tup)
+            tensor_op = str_to_op(op_name)
+            # append in output_dict
+            output_dict[op_name] = SymbolicHamiltonian(tensor_op).dense.matrix
+    return output_dict
+
+
+def str_to_op(name: str):
+    tensor_op = 1
```
Comment (andrea-pasquale) [inline src/qibo/models/dbi/utils.py:55]: I would suggest to add some docstrings for this function. Even if it is already pretty clear.

File: src/qibo/models/dbi/utils.py (hunk 1 above)
Comment (Edoardo-Pedicillo) [reply src/qibo/models/dbi/utils.py:125]: > Please correct me if I understood wrongly... so we deepcopy dbi_object in the function and update it by default?

You don't need to deeply copy `dbi_object`, you could just update its mode.

> To run the double bracket rotation, the user will do dbi_input, idx_max_loss, step_optimal = select_best_dbr_generator(dbi_input)?

I suggested something like `dbi_input, norms_off_diagonal_restriction, step_optimal = select_best_dbr_generator(dbi_input, ...)`. In this way you can delete `select_best_dbr_generator_and_run` and if you want to run the best dbi use `dbi_input(step_optimal)`.

File: src/qibo/models/dbi/utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Edoardo-Pedicillo) [reply src/qibo/models/dbi/utils.py:61]: Ok, since is not defined somewhere, you could specify it in the docs. I suggest renaming the function `best_dbr_generator`, because it is implicit that you are selecting, if you are looking for the best value.

File: src/qibo/models/dbi/utils.py (hunk 1 above)
Comment (andrea-pasquale) [reply src/qibo/models/dbi/utils.py:125]: Actually to further simplify this function can we consider using `hyperopt` directly? At the end we are doing something similar to `hyperopt_step` with the only difference that we are optimizing over a larger space which in this case consist of `(steps, ds, mode)`. I can come up with a prototype.

**Review by andrea-pasquale — COMMENTED** [review]
Thanks @Sam-XiaoyueLi for all the updates.
Could you please add also some tests? In Qibo we use [pytest](https://docs.pytest.org/en/7.4.x/). You just need to add some functions in this [file](https://github.com/qiboteam/qibo/blob/dbf_migrate/tests/test_models_dbi.py) that test the new features that you are adding.

File: src/qibo/models/dbi/utils.py (hunk 1 above)
Comment (Sam-XiaoyueLi) [reply src/qibo/models/dbi/utils.py:134]: Thanks a lot Andrea! I have tried the code and made simple modifications that allow us to store the losses and update h etc.. However, I have an issue with it being very unstable, in a sense that the off-diagonal norm does not always decrease.
The function:
```python
def hyperopt_step_mode(
                        ham: hamiltonians.Hamiltonian,
                        d: list,
                        step_min: float = 1e-5,
                        step_max: float = 1,
                        max_evals: int = 100,
                        verbose: bool = False):
    # Define the search space
    space = {
        "step": hyperopt.hp.uniform("step", step_min, step_max),
        "mode": hyperopt.hp.choice("mode", [
            {"generator": DoubleBracketGeneratorType.canonical},
            {"generator": DoubleBracketGeneratorType.single_commutator, 
             "d_index": hyperopt.hp.choice("d", list(range(len(d))))}  # Using index for 'd'
        ])
    }

    # Define the loss function
    def loss(args):
        config, step = args["mode"], args["step"]
        dbi = DoubleBracketIteration(hamiltonian=ham, mode=config["generator"])
        if config["generator"] is DoubleBracketGeneratorType.canonical:
            return dbi.loss(step=step)
        else:
            d_index = config["d_index"]
            return dbi.loss(step=step, d=d[d_index])  # Use the index to get the value of 'd'

    # Perform the optimization
    best = hyperopt.fmin(fn=loss,
                         space=space,
                         algo=hyperopt.tpe.suggest,
                         max_evals=max_evals,
                         verbose=verbose)

    # Evaluate the best hyperparameters
    best_evaluated = hyperopt.space_eval(space, best)
    best_loss = loss(best_evaluated)
    output = {
        "mode": best_evaluated["mode"]["generator"],
        "step": best_evaluated["step"],
        "d": d[best_evaluated["mode"]["d_index"]] if "d_index" in best_evaluated["mode"] else None,
        "d_index": best_evaluated["mode"]["d_index"] if "d_index" in best_evaluated["mode"] else None,
        "loss": best_loss,
    }

    return output
```
Test code:
```python
NSTEPS = 15
max_evals = 100
step_max = 1
op_names = Z_names + ['Canonical']
# add in initial values for plotting
off_diagonal_norm_history = [dbi.off_diagonal_norm]
steps = [0]
ham = H_TFIM
dbi_hyper_pauli = DoubleBracketIteration(ham, mode=DoubleBracketGeneratorType.single_commutator)
for _ in range(NSTEPS):
    output = hyperopt_step_mode(dbi_hyper_pauli.h, Z_ops)
    if output['mode'] == DoubleBracketGeneratorType.canonical:
        op_ind = -1
        dbi_hyper_pauli(output['step'], mode=DoubleBracketGeneratorType.canonical)
    else:
        op_ind = output['d_index']
        dbi_hyper_pauli(output['step'], d=output['d'], mode=DoubleBracketGeneratorType.single_commutator)
    print(f"New optimized step at iteration {_+1}/{NSTEPS}: {output['step']} with operator {op_names[op_ind]}, loss {dbi_hyper_pauli.off_diagonal_norm}")
```

Comment (andrea-pasquale) [reply src/qibo/models/dbi/utils.py:133]: Thanks @Sam-XiaoyueLi. I will have a look. It would be interesting to understand more why `hyperopt` is so unstable

Comment (marekgluza) [reply src/qibo/models/dbi/utils.py:133]: Summary of discussion with Sam:
- often the optimization landscape has spikes
- Sam made some observations qualitatively when hyperopt gets stuck or not, it makes a difference if there are several spikey minima of same depth or not
- keep in mind we are using quite few function evaluations ~100
- I think ultimately we will take scheduling formulas from Andrew, get an initial guess for the apex of the parabola and narrow down the hyperopt window
- the hyperopt window for the first DBI step should be quite large, around 0.3 for what I know
- I added to the timeline the task to do some scaling on the apex of the parabola for the first minimum as the number of qubits increases @wrightjandrew @Sam-XiaoyueLi you could work on this together and then later check with @taufiqmurtadho using tenpy
- we might add a scheduling strategy to have the first hyperopt step extensive, with a large window; then subsequently narrow it down

**Review by marekgluza — APPROVED** [review]

**Review by marekgluza — APPROVED** [review]
thanks Sam

File: tests/test_models_dbi_utils.py [hunk 4]
```diff
@@ -42,13 +42,13 @@ def test_select_best_dbr_generator_and_run(backend, nqubits, step):
     initial_off_diagonal_norm = dbi.off_diagonal_norm
 
     for _ in range(NSTEPS):
-        idx, step_optimize = select_best_dbr_generator_and_run(
+        idx, step_optimize, flip_sign = select_best_dbr_generator_and_run(
```
Comment (marekgluza) [inline tests/test_models_dbi_utils.py:45]: @Edoardo-Pedicillo When we have [D,H] then sometimes D-> (-D) is needed. Sam is working here with exploring all idx and finds this sign by the CS_angle function.

What is better for runtime: passing idx and then building D again by sign * "Pauli_ops[idx]" or passing back the operator e.g.
`D,s  =  select_best_dbr_generator_and_run(...)`?

File: tests/test_models_dbi_utils.py [hunk 5]
```diff
@@ -42,13 +42,13 @@ def test_select_best_dbr_generator_and_run(backend, nqubits, step):
...
             dbi, Z_ops, step=step, compare_canonical=True
         )
         if idx == len(Z_ops):
             dbi(step=step_optimize, mode=DoubleBracketGeneratorType.canonical)
             dbi.mode = DoubleBracketGeneratorType.single_commutator
         else:
-            dbi(step=step_optimize, d=Z_ops[idx])
+            dbi(step=step_optimize, d=flip_sign * Z_ops[idx])
```
Comment (marekgluza) [inline tests/test_models_dbi_utils.py:52]: @andrea-pasquale @MatteoRobbiati @Edoardo-Pedicillo 

This is a key change which implements for the strategy of using $D=Z_1^{idx_0}\ldots Z_L^{idx_L}$ the theoretical discussion roughly saing for DBI either +D or -D will be the right one. Here Sam implemented it by checking the CS angle.

In general, we can now keep an eye on this for other strategies and basically check if $\langle W_{test}, W_{canonical}\rangle>0$, otherwise flip. 

Thanks @Sam-XiaoyueLi

**Review by marekgluza — APPROVED** [review]

File: tests/test_models_dbi_utils.py (hunk 4 above)
Comment (Edoardo-Pedicillo) [reply tests/test_models_dbi_utils.py:45]: As suggested here https://github.com/qiboteam/qibo/pull/1143#discussion_r1447015892 `select_best_dbr_generator_and_run` should return the `dbi_object` with the best mode (i.e., with the best operator), `step_optimal` and `norms_off_diagonal_restriction`.

> What is better for runtime: passing idx and then building D again by sign * "Pauli_ops[idx]" or passing back the operator e.g.
D,s  =  select_best_dbr_generator_and_run(...)?

Since at the end you should always build D, I think the difference is negligible.

**Comment by Edoardo-Pedicillo** [comment]
@Sam-XiaoyueLi feel free to close the comments when you think they are outdated (press the `Resolve conversation` button)

When you are ready for a new review, ask it pressing the circle icon with the two arrows.

File: src/qibo/models/dbi/double_bracket.py [hunk 6]
```diff
@@ -103,12 +102,12 @@ def off_diag_h(self):
 
     @property
     def off_diagonal_norm(self):
-        """Norm of off-diagonal part of H matrix."""
+        r"""Hilbert Schmidt norm of off-diagonal part of H matrix: \sqrt{A^\dag A}"""
```
Comment (marekgluza) [inline src/qibo/models/dbi/double_bracket.py:105]: Trace is missing in the doc string:

`r"""Hilbert Schmidt norm of off-diagonal part of H matrix: \sqrt{Tr[A^\dag A]}"""`

File: src/qibo/models/dbi/utils.py [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+from qibo.config import raise_error
+from qibo.hamiltonians import Hamiltonian, SymbolicHamiltonian
+from qibo.models.dbi.double_bracket import (
+    DoubleBracketGeneratorType,
+    DoubleBracketIteration,
+)
+
+
+def generate_Z_operators(nqubits: int):
+    """Generate a dictionary containing 1) all possible products of Pauli Z operators for L = n_qubits and 2) their respective names.
```
Comment (marekgluza) [inline src/qibo/models/dbi/utils.py:18]: Might want to say what is the type of the values in the dict, SymbolicHamiltonian or?

File: src/qibo/models/dbi/utils.py [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+    Returns:
+        The updated dbi_object, index of the optimal diagonal operator, respective step duration, and evolution direction.
+    """
+    norms_off_diagonal_restriction = [0 for i in range(len(d_list))]
+    optimal_steps = [0 for i in range(len(d_list))]
+    flip_list = [1 for i in range(len(d_list))]
+    for i, d in enumerate(d_list):
+        # prescribed step durations
+        dbi_eval = deepcopy(dbi_object)
+        flip_list[i] = CS_angle_sgn(dbi_eval, d)
```
Comment (marekgluza) [inline src/qibo/models/dbi/utils.py:102]: :ok_hand:

**Review by marekgluza — CHANGES_REQUESTED** [review]
Requesting change in the docstring as proposed in the comment.

Optional suggestion to specify what the Z operators dictionary contains 

Otherwise, wonderful. This PR is very instructive and we should think what to take over and what to design for the new strategies ahead.

Thanks for all the hard work and if it will be needed please block time to prepare changes for the merge into the main branch so that you will not feel under strain with other tasks.

:+1: :+1:  :):)

**Review by marekgluza — APPROVED** [review]

**Review by marekgluza — APPROVED** [review]

**Review by marekgluza — APPROVED** [review]
Sam's explanation:
The list is used to compute changes in the off-diagonal restriction.
Previously if the list had a 0 entry then it was possible that the 'default' entry with a 0 would have been selected because dropping to 0 would 'mean' we get a diagonalized matrix in that step. (Full diagonalization in 1 step is not possible because of Galois theory)

Now, in the cases where there would be a 0 remaining in the list (impossible to move the dbi forward with that choice of operator) there is correctly no change in the off-diagonal norm.

**Comment by marekgluza** [comment]
> Thanks @Sam-XiaoyueLi for adding these interesting features. You can find below some comments. Let me know if you need help with any of the points that I raised.

Andrea, how does it work with merging in this situation? Should we request for your approval or 'somehow' go ahead with merge?

**Comment by andrea-pasquale** [comment]
> > Thanks @Sam-XiaoyueLi for adding these interesting features. You can find below some comments. Let me know if you need help with any of the points that I raised.
> 
> Andrea, how does it work with merging in this situation? Should we request for your approval or 'somehow' go ahead with merge?

In order to merge the pull request two approvals are needed. I need to check the last changes and then I can approve.

File: src/qibo/models/dbi/utils.py [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+        optimal_steps.append(step_best)
+        norms_off_diagonal_restriction.append(dbi_eval.off_diagonal_norm)
+    # find best d
+    idx_max_loss = norms_off_diagonal_restriction.index(
+        min(norms_off_diagonal_restriction)
+    )
+    flip = flip_list[idx_max_loss]
+    step_optimal = optimal_steps[idx_max_loss]
+    dbi_eval = deepcopy(dbi_object)
+    if idx_max_loss == len(d_list) and compare_canonical is True:
```
Comment (andrea-pasquale) [inline src/qibo/models/dbi/utils.py:145]: Here I guess that you could just check if `d` is $\Delta(H)$, right? 
It should improve readability

File: src/qibo/models/dbi/utils.py [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+        # canonical
+        dbi_eval(step=step_optimal, mode=DoubleBracketGeneratorType.canonical)
+
+    else:
+        d_optimal = flip * d_list[idx_max_loss]
+        dbi_eval(step=step_optimal, d=d_optimal)
+    return dbi_eval, idx_max_loss, step_optimal, flip
```
Comment (andrea-pasquale) [inline src/qibo/models/dbi/utils.py:152]: Do we really need `idx_max_loss` and `flip`? We could just return `step_optimal` and `best_d`, right?

File: src/qibo/models/dbi/utils.py [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+    """
+    norms_off_diagonal_restriction = [
+        dbi_object.off_diagonal_norm for _ in range(len(d_list))
+    ]
+    optimal_steps = [0 for _ in range(len(d_list))]
+    flip_list = [1 for _ in range(len(d_list))]
+    for i, d in enumerate(d_list):
+        # prescribed step durations
+        dbi_eval = deepcopy(dbi_object)
+        flip_list[i] = CS_angle_sgn(dbi_eval, d)
```
Comment (MatteoRobbiati) [inline src/qibo/models/dbi/utils.py:104]:
```suggestion
        flip_list[i] = cs_angle_sgn(dbi_eval, d)
```

File: src/qibo/models/dbi/utils.py [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+    Args:
+        dbi_object (_DoubleBracketIteration): The target DoubleBracketIteration object.
+        d_list (list): List of diagonal operators (np.array) to run from.
+        step (float): Fixed iteration duration.
+            Defaults to ``None``, uses hyperopt.
+        step_min (float): Minimally allowed iteration duration.
+        step_max (float): Maximally allowed iteration duration.
+        max_evals (int): Maximally allowed number of evaluation in hyperopt.
+        compare_canonical (bool): If `True`, the optimal diagonal operator chosen from "d_list" is compared with the canonical bracket.
+        mode (_DoubleBracketGeneratorType): DBI generator type used for the selection.
```
Comment (MatteoRobbiati) [inline src/qibo/models/dbi/utils.py:91]:
```suggestion
        dbi_object (`DoubleBracketIteration`): the target DoubleBracketIteration object.
        d_list (list): list of diagonal operators (np.array) to run from.
        step (float): fixed iteration duration.
            Defaults to ``None``, uses hyperopt.
        step_min (float): minimally allowed iteration duration.
        step_max (float): maximally allowed iteration duration.
        max_evals (int): maximally allowed number of evaluation in hyperopt.
        compare_canonical (bool): if `True`, the optimal diagonal operator chosen from "d_list" is compared with the canonical bracket.
        mode (`DoubleBracketGeneratorType`): DBI generator type used for the selection.
```

File: src/qibo/models/dbi/utils.py [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+        dbi_eval(step=step_optimal, mode=DoubleBracketGeneratorType.canonical)
+
+    else:
+        d_optimal = flip * d_list[idx_max_loss]
+        dbi_eval(step=step_optimal, d=d_optimal)
+    return dbi_eval, idx_max_loss, step_optimal, flip
+
+
+def cs_angle_sgn(dbi_object, d):
+    """Calculates the sign of Cauchy-Schwarz Angle $$<W(Z), W(canonical)>_{HS}$$"""
```
Comment (MatteoRobbiati) [inline src/qibo/models/dbi/utils.py:156]:
```suggestion
    """Calculates the sign of Cauchy-Schwarz Angle :math:`\\langle W(Z), W({\\rm canonical}) \\rangle_{\\rm HS}`."""
```

File: tests/test_models_dbi_utils.py [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+from qibo.hamiltonians import Hamiltonian
+from qibo.models.dbi.double_bracket import (
+    DoubleBracketGeneratorType,
+    DoubleBracketIteration,
+)
+from qibo.models.dbi.utils import *
+from qibo.quantum_info import random_hermitian
+
+NSTEPS = 5
+"""Number of steps for evolution."""
```
Comment (MatteoRobbiati) [inline tests/test_models_dbi_utils.py:16]:
```suggestion
```

File: tests/test_models_dbi_utils.py [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+
+
+@pytest.mark.parametrize("nqubits", [3, 4, 5])
+def test_generate_Z_operators(nqubits):
```
Comment (MatteoRobbiati) [inline tests/test_models_dbi_utils.py:20]:
```suggestion
def test_generate_Z_operators(backend, nqubits):
```

Comment (MatteoRobbiati) [reply tests/test_models_dbi_utils.py:20]: This can be useful anyway to test all the backends.

**Review by MatteoRobbiati — CHANGES_REQUESTED** [review]
Thanks a lot @Sam-XiaoyueLi for this. 

Some suggestions follow. Many of them are very small comments related to the docstrings. 
Again on docstrings: I think we should add the `utils` file to the documentation. I can do this once the last fixes are done.

The tests are failing because the variable `CS_angle_sgn` is not defined. I think you wanted to call `cs_angle_sgn`. This will be fixed adding one of my following suggestions.

File: src/qibo/models/dbi/utils.py
[the remark this answers is not in the document]
Comment (marekgluza) [reply src/qibo/models/dbi/utils.py:1]: Hi @MatteoRobbiati thanks that would be helpful! There's Chinese New Year now so we are both off next week but once @Sam-XiaoyueLi is back it would be great to close the PR! :)

[the remark this answers is not in the document]
Comment (marekgluza) [reply src/qibo/models/dbi/utils.py:1]: that would be fantastic if you could merge after your changes, thanks @MatteoRobbiati 

if there are any issues, you can raise separate issues too and assign @Sam-XiaoyueLi  and then we can split the requests into smaller merges. Will be easier to close the PRs? Happy CNY! :)

**Comment by MatteoRobbiati** [comment]
I did some modifications. 
Only a couple of @andrea-pasquale's comments are missing. 
To me this is almost ready to go.

**Review by andrea-pasquale — APPROVED** [review]
